File: code/socket2/posix/daytimetcpcli.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <netinet/in.h>

#define MAXLINE 1024
#define SA struct sockaddr

int main(int argc, char **argv)
{
    int                 sockfd, n;
    int                 port;
    char                recvline[MAXLINE + 1];
    struct sockaddr_in  servaddr;

    if (argc != 3) {
        fprintf(stderr, "usage: %s <IPaddress> <port>\n", argv[0]);
        exit(1);
    }

    port = atoi(argv[2]);
    if (port <= 0 || port > 65535) {
        fprintf(stderr, "invalid port: %s\n", argv[2]);
        exit(1);
    }

    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
        perror("socket error");
        exit(1);
    }

    memset(&servaddr, 0, sizeof(servaddr));
    servaddr.sin_family = AF_INET;
    servaddr.sin_port   = htons((uint16_t) port);   /* port from argv */
    if (inet_pton(AF_INET, argv[1], &servaddr.sin_addr) <= 0) {
        fprintf(stderr, "inet_pton error for %s\n", argv[1]);
        exit(1);
    }

    if (connect(sockfd, (SA *) &servaddr, sizeof(servaddr)) < 0) {
        perror("connect error");
        exit(1);
    }

    while ((n = read(sockfd, recvline, MAXLINE)) > 0) {
        recvline[n] = 0;    /* null terminate */
        if (fputs(recvline, stdout) == EOF) {
            perror("fputs error");
            exit(1);
        }
    }
    if (n < 0) {
        perror("read error");
        exit(1);
    }

    exit(0);
}

File: code/socket2/python/daytimetcp/daytimetcpcli.py
import socket
import sys

def main():
    if len(sys.argv) != 3:
        sys.exit("usage: a.out <IPaddress> <port>")

    host = sys.argv[1]
    try:
        port = int(sys.argv[2])
    except ValueError:
        sys.exit(f"invalid port: {sys.argv[2]}")

    try:
        sockfd = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    except OSError as e:
        sys.exit(f"socket error: {e}")

    servaddr = (host, port)

    try:
        sockfd.connect(servaddr)
    except OSError as e:
        sys.exit(f"connect error: {e}")

    try:
        while True:
            data = sockfd.recv(1024)
            if not data:
                break
            sys.stdout.write(data.decode(errors="replace"))
            sys.stdout.flush()
    except OSError as e:
        sys.exit(f"read error: {e}")
    finally:
        sockfd.close()

if __name__ == "__main__":
    main()
